Source/Scene/Cesium3DTileBatchTable.js

```diff
@@ -341,32 +471,266 @@ define([
+        while (true) {
+            var result = endConditionCallback(hierarchy, instanceIndex);
+            if (defined(result)) {
+                // The end condition was met, stop the traversal and return the result
+                return result;
+            }
+            var parentId = hierarchy.parentIds[instanceIndex];
+            if (parentId === instanceIndex) {
+                // Stop the traversal when the instance has no parent (its parentId equals itself)
+                break;
+            }
+            instanceIndex = parentId;
+        }
+    }
+
+    function traverseHierarchy(hierarchy, instanceIndex, endConditionCallback) {
+        // Traverse over the hierarchy and process each instance with the endConditionCallback.
+        // When the endConditionCallback returns a value, the traversal stops and that value is returned.
+        var parentCounts = hierarchy.parentCounts;
+        if (defined(parentCounts)) {
+            return traverseHierarchyMultipleParents(hierarchy, instanceIndex, endConditionCallback);
+        }
+        return traverseHierarchySingleParent(hierarchy, instanceIndex, endConditionCallback);
+    }
+
+    function hasPropertyInHierarchy(batchTable, batchId, name) {
+        var hierarchy = batchTable._batchTableHierarchy;
+        var result = traverseHierarchy(hierarchy, batchId, function(hierarchy, instanceIndex) {
+            var classId = hierarchy.classIds[instanceIndex];
+            var instances = hierarchy.classes[classId].instances;
+            if (defined(instances[name])) {
+                return true;
+            }
+        });
+        return defined(result);
+    }
+
+    function getPropertyNamesInHierarchy(batchTable, batchId, names) {
+        var hierarchy = batchTable._batchTableHierarchy;
+        traverseHierarchy(hierarchy, batchId, function(hierarchy, instanceIndex) {
+            var classId = hierarchy.classIds[instanceIndex];
+            var instances = hierarchy.classes[classId].instances;
+            for (var name in instances) {
+                if (instances.hasOwnProperty(name)) {
+                    names[name] = true;
+                }
+            }
+        });
+    }
+
+    function getHierarchyProperty(batchTable, batchId, name) {
+        var hierarchy = batchTable._batchTableHierarchy;
+        return traverseHierarchy(hierarchy, batchId, function(hierarchy, instanceIndex) {
+            var classId = hierarchy.classIds[instanceIndex];
+            var instanceClass = hierarchy.classes[classId];
+            var indexInClass = hierarchy.classIndexes[instanceIndex];
+            var propertyValues = instanceClass.instances[name];
+            if (defined(propertyValues)) {
+                if (defined(propertyValues.typedArray)) {
+                    return getBinaryProperty(propertyValues, indexInClass);
+                }
+                return clone(propertyValues[indexInClass], true);
```

Comment:
I modeled this after [getProperty](https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Source/Scene/Cesium3DTileBatchTable.js#L405). I believe the idea is that we don't want to make batch table properties that are objects editable via the `getProperty` function.